idx/memory/tag_query.go

```diff
@@ -653,30 +699,27 @@ func (q *TagQuery) testByTagMatch(def *idx.Archive) bool {
 		}
 	}
 
-	for _, tag := range def.Tags {
-		equal := strings.Index(tag, "=")
-		if equal < 0 {
-			corruptIndex.Inc()
-			log.Errorf("memory-idx: ID %q has tag %q in index without '=' sign", def.Id, tag)
+	for _, tag := range def.Tags.KeyValues {
+		if _, ok := q.tagMatch.missCache.Load(tag.Key); ok {
 			continue
 		}
-		key := tag[:equal]
 
-		if _, ok := q.tagMatch.missCache.Load(key); ok {
-			continue
+		key, err := idx.IdxIntern.GetStringFromPtr(tag.Key)
+		if err != nil {
+			log.Error("memory-idx: Failed to retrieve uintptr for interned tag key: ", err)
+			internError.Inc()
```

Comment:
at this point, better `continue` to the next tag

Reply:
Or should we panic?

Reply:
I don't think we want to panic, dealing with it an continuing somehow is better. "dealing with it" could mean either directly return with an error, or just increase the `internError` metric and continue as good as possible. but panicking can be risky as in some code paths locks might be left open because nobody expected a panic, or stuff like that.

Reply:
see number 11 at https://github.com/grafana/metrictank/blob/master/docs/CONTRIBUTING.md#when-contributing-prs

Reply:
I thought we decided that interning errors were unrecoverable, at least in other parts of this code.

Reply:
I think it depends:
* if the consequence of an interning error is that we error a response for a render query, or we return data to the user that has a tag key or value that can't be set properly (but instead we can set it to something like `intern-error`) then i would classify that as recoverable
* if the consequence is that data may be corrupted in the index, in the storage etc, then it is unrecoverable.

> panicking can be risky as in some code paths locks might be left open because nobody expected a panic, or stuff like that.

I don't agree with this. panic is the way to shut down the process in case of an unrecoverable error. i am not aware of any reason why this wouldn't work, and if there was such a reason, we should fix it, because our policy is to panic upon unrecoverable errors.